File: sdk/ruby/lib/anyvali/schemas/never_schema.rb
# frozen_string_literal: true

module AnyVali
  class NeverSchema < Schema
    def initialize(**kwargs)
      super(kind: "never", **kwargs)
    end

    protected

    def validate(value, path, issues, context)
      issues << ValidationIssue.new(
        code: IssueCodes::INVALID_TYPE,
        path: path,
        expected: "never",
        received: Schema.type_name(value)
      )
    end
  end
end

File: sdk/csharp/src/AnyVali/Schemas/NeverSchema.cs
namespace AnyVali.Schemas;

public sealed class NeverSchema : Schema<object?>
{
    internal override object? Validate(object? input, ValidationContext ctx)
    {
        ctx.Issues.Add(new ValidationIssue
        {
            Code = IssueCodes.InvalidType,
            Message = "Expected never (no value is valid)",
            Path = ctx.ClonePath(),
            Expected = "never",
            Received = DescribeType(input),
        });
        return null;
    }

    internal override Dictionary<string, object?> ToNode() =>
        new() { ["kind"] = "never" };

    internal override Schema Clone() => new NeverSchema
    {
        DefaultValue = DefaultValue, CoercionCfg = CoercionCfg, IsPortable = IsPortable,
    };
}
